pkg/virt-handler/realtime.go

```diff
@@ -0,0 +1,202 @@
+	enabled    maskType = true
+	disabled   maskType = false
+)
+
+var (
+	// parse CPU Mask expressions
+	cpuRangeRegex  = regexp.MustCompile(`^(\d+)-(\d+)$`)
+	negateCPURegex = regexp.MustCompile(`^\^(\d+)$`)
+	singleCPURegex = regexp.MustCompile(`^(\d+)$`)
+
+	// parse thread comm value expression
+	vcpuRegex = regexp.MustCompile(`^CPU (\d+)/KVM\n$`) // These threads follow this naming pattern as their command value (/proc/{pid}/task/{taskid}/comm)
+// QEMU uses threads to represent vCPUs.
+
+)
+
+// configureRealTimeVCPUs parses the realtime mask value and configured the selected vcpus
+// for real time workloads by setting the scheduler to FIFO and process priority equal to 1.
+func (d *VirtualMachineController) configureVCPUScheduler(vmi *v1.VirtualMachineInstance) error {
+	res, err := d.podIsolationDetector.Detect(vmi)
+	if err != nil {
+		return err
+	}
+	qemuProcess, err := isolation.GetQEMUProcess(res.PPid())
+	if err != nil {
+		return err
+	}
+
+	vcpus, err := getVCPUThreadIDs(qemuProcess.Pid())
+	if err != nil {
+		return err
+	}
+	mask, err := parseCPUMask(vmi.Spec.Domain.CPU.Realtime.Mask)
+	if err != nil {
+		return err
+	}
+	for vcpuID, threadID := range vcpus {
+
+		if isRealtimeVCPU(mask, vcpuID) {
+			param := SchedParam{sched_priority: 1}
+			tid, err := strconv.Atoi(threadID)
+			if err != nil {
+				return err
+			}
+			err = schedSetScheduler(tid, SCHED_FIFO, param)
+			if err != nil {
+				return fmt.Errorf("failed to set FIFO scheduling and priority 1 for thread %d: %w", tid, err)
+			}
+		}
+	}
+	return nil
+}
+
+func isRealtimeVCPU(parsedMask map[string]maskType, vcpuID string) bool {
+	if len(parsedMask) == 0 {
+		return true
+	}
+	if t, ok := parsedMask[vcpuID]; ok {
+		return t == enabled
+	}
+	return false
+}
+
+func isVCPU(comm []byte) (string, bool) {
+	if !vcpuRegex.MatchString(string(comm)) {
+		return "", false
+	}
+	v := vcpuRegex.FindSubmatch(comm)
+	return string(v[1]), true
+}
+func getVCPUThreadIDs(pid int) (map[string]string, error) {
+
+	p := filepath.Join(string(os.PathSeparator), "proc", strconv.Itoa(pid), "task")
+	d, err := os.ReadDir(p)
+	if err != nil {
+		return nil, err
+	}
+	ret := map[string]string{}
+	for _, f := range d {
+		if f.IsDir() {
+			c, err := os.ReadFile(filepath.Join(p, f.Name(), "comm"))
+			if err != nil {
+				return nil, err
+			}
+			if v, ok := isVCPU(c); ok {
+				ret[v] = f.Name()
+			}
+		}
+	}
+	return ret, nil
+}
+
+// parseCPUMask parses the mask and maps the results into a structure that contains which
+// CPUs are enabled or disabled for the scheduling and priority changes.
+// This implementation duplicates the libvirt parsing logic defined here:
+// https://github.com/libvirt/libvirt/blob/56de80cb793aa7aedc45572f8b6ec3fc32c99309/src/util/virbitmap.c#L382
+// except that in this case it uses a map[string]maskType instead of a bit array.
+func parseCPUMask(mask string) (map[string]maskType, error) {
+
+	if len(strings.TrimSpace(mask)) == 0 {
```

Comment:
nit:
```suggestion
	if strings.TrimSpace(mask) != "" {
```

Reply:
Looking at the code I should return an empty map if the mask is either empty or full of spaces. Only return error if the parsing fails.
```
	vcpus := make(map[string]maskType)
	if strings.TrimSpace(mask) == "" {
		return vcpus, nil
	}
```
The reason is that `Mask` is a string and will never be nil, so it should always treat the empty case as default.

Reply:
great, thanks for the suggestion! :smile:

Reply:
After further thought I think a mask filled with spaces should be considered invalid, compared with an empty string.
```go
	if len(mask) == 0 {
		return nil, nil
	}
```
WDYT?